src/java/org/apache/cassandra/service/accord/AccordStateCache.java

```diff
@@ -51,71 +55,107 @@
 {
     private static final Logger logger = LoggerFactory.getLogger(AccordStateCache.class);
 
-    private static class WriteOnlyGroup<K, V extends AccordState<K>>
+    public interface LoadFunction<K, V>
     {
-        private boolean locked = false;
-        private List<AccordState.WriteOnly<K, V>> items = new ArrayList<>();
+        AsyncResult<Void> apply(K key, Consumer<V> valueConsumer);
```

Comment:
in the spirit of chains vs result, I think this should be

```
AsyncChain<V> apply(K key);
```

Looking at `PendingLoad` it looks like we are bad about handling the exception case...  For example look at `org.apache.cassandra.service.accord.AccordStateCache.Node#maybeFinishLoad` it just sets the `state = null` when we see an error!  This puts us in a confusing state where `isLoaded() == true`, but we are not loaded!  Which I guess is why `value()` has to have this check

```
Invariants.checkState(isLoaded() && state != null);
```

I tried working on a patch to clean this up and found issues with `AsyncResult` (need `getCauseNow`, `getNow` methods), and `AsyncResults` (`reduce` doesn't handle types properly, causing `List<AsyncChain<?>>` to not work)


For `accord.utils.async.AsyncResults#reduce` it should be

```
<T> AsyncChain<V> reduce(Collection<? extends AsyncChain<? extends V>> results, BinaryOperator<V> reducer)
```

Reply:
We need an AsyncResult in this case because we need to check if the operation has completed.

Reply:
we need result *after* the apply, we don't need it before.  The current logic is

* before complete, call `consumer`
* save result to a `AsyncResult`

We are reinventing call backs, so rather than push it in, if we return a `AsyncChain` it has the same semantics as

```
apply(key)
                 .accept(r -> this.value = r)
                 .beginAsResult()
```

> Looking at PendingLoad it looks like we are bad about handling the exception case

This is also another reason to switch to `AsyncChain` as the API is supposed to make us think about failures, so by trying to create custom callbacks in the execution we ignore errors, leading to us putting the cache in a bad state that later gets rejected (due to value being null)